File: haskell/glue/src/Glue/Lib/Bool/Or.hs
module Glue.Lib.Bool.Or where

import Glue.Eval (Eval, eval, throwError)
import Glue.Eval.Exception (wrongNumberOfArguments)
import Glue.IR (IR (..), isTruthy)

or_ :: IR Eval
or_ = Special orImpl

orImpl :: [IR Eval] -> Eval (IR Eval)
orImpl [a, b] = do
    a' <- eval a
    if isTruthy a'
        then pure (Bool True)
        else Bool . isTruthy <$> eval b
orImpl _ = throwError wrongNumberOfArguments

File: dart/glue/lib/src/lib/bool/or.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

final Ir or_ = IrSpecial(orImpl);

Eval<Ir> orImpl(List<Ir> args) {
  return switch (args) {
    [final a, final b] => eval(a).bind((a_) {
      return isTruthy(a_)
          ? Eval.pure(IrBool(true))
          : eval(b).bind((v) => Eval.pure(IrBool(isTruthy(v))));
    }),
    _ => throwError(wrongNumberOfArguments()),
  };
}
